### PR TITLE
Improve less-than comparison between arrays

### LINKED ISSUE #39299: Less than comparisons between arrays and sorting arrays of arrays is unpredictable
**Godot version:**
3.2.1.stable

**OS/device including version:**
Windows10

**Issue description:**
I just tried to sort an array of pairs of numbers `[[2.061036, 1], [2.599602, 100], [-0.833468, 1]]` and the answer wasn't right.  

It appears that Array objects are not compared by the value of their contents, even when they are arrays of one element, as shown in the example below.  It is not obvious what is being compared.  Object pointer addresses?  

Sorting arrays of tuples is a technique I use a lot in C++ to avoid the hassle of having to define a comparison function pointer required by `sort_custom()`.  Python used to have a similar `cmp(a, b)` function for its sort feature, but it was [taken out](https://portingguide.readthedocs.io/en/latest/comparisons.html#the-cmp-argument) of Python3 in favour of a simpler `key(a)` function.

The [class_array.html documentation](https://docs.godotengine.org/en/3.2/classes/class_array.html#class-array-method-sort-custom) gives an example use of `sort_custom()` which would have worked with a normal `sort()` in Python.  If this is a message about how to do it, it's probably too subtle.  There is a warning about sorting strings containing numbers, but no warning that sorting arrays of arrays is going to be a problem.  

If comparing Arrays doesn't work in the way expected, then it would be better if it gave an Invalid Operation Type error (as it does when you compare to Dictionaries) than to allow it to be written.

**Steps to reproduce:**
```Python
	print("0.8 < 2.5: ", 0.8 < 2.5)          # ---> True
	print("[0.8] < [2.5]: ", [0.8] < [2.5])  # ---> False
	print("[9.8] < [2.5]: ", [9.8] < [2.5])  # ---> True
```

**Minimal reproduction project:**

I've put the above code into the run() function of an editorscript .gd file and used Control-Shift-X to run it:

```
tool
extends EditorScript
func _run():
    ...
```

### PR DESCRIPTION
**Pull request by RexMaHaoXuan**
It appears that the issue is simply caused by the misuse of element dereference.
Also included some relevant test cases.

*Bugsquad edit: This closes #39299.*

### CODE REVIEW

**Comment by Calinou** [comment]
Thanks for opening a pull request!

Please [amend the commit to use a human-readable commit message](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#format-your-commit-messages-with-readability-in-mind). This way, people reading the Git log can easily figure out what's going on.

**Comment by akien-mga** [comment]
Tests should go in the `tests` folder.

File: tests/test_array.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+#include "core/variant/array.h"
+#include "core/string/print_string.h"
+#include "tests/test_macros.h"
+
+#include "thirdparty/doctest/doctest.h"
+
+namespace TestArray {
+TEST_CASE("Array < Single Element") {
+	const Array a1 = [0.8]
+	const Array a2 = [9.5]
```
Comment (Calinou) [inline tests/test_array.h:43]:
```suggestion
	const Array a1 = [0.8];
	const Array a2 = [9.5];
```

File: tests/test_array.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+	CHECK_MESSAGE(
+			a1 < a2,
+			"Array with a smaller item at index 0 should be smaller");
+}
+
+TEST_CASE("Array < More Elements") {
+	const Array a1 = [1, 1]
+	const Array a2 = [1, 2]
```
Comment (Calinou) [inline tests/test_array.h:52]:
```suggestion
	const Array a1 = [1, 1];
	const Array a2 = [1, 2];
```

File: tests/test_array.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+	CHECK_MESSAGE(
+			a1 < a2,
+			"Array with a smaller item at index 1 should be smaller");
+}
+
+TEST_CASE("Array < More Elements 2") {
+	const Array a1 = [1, 2, 3]
+	const Array a2 = [2, 3, 4]
```
Comment (Calinou) [inline tests/test_array.h:61]:
```suggestion
	const Array a1 = [1, 2, 3];
	const Array a2 = [2, 3, 4];
```

File: tests/test_array.h [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+	CHECK_MESSAGE(
+			a1 < a2,
+			"Array with a smaller item at index 1 should be smaller");
+}
+
+TEST_CASE("Array < More Elements 3") {
+	const Array a1 = ["b", 2, 3]
+	const Array a2 = ["a", 2, 3]
```
Comment (Calinou) [inline tests/test_array.h:70]:
```suggestion
	const Array a1 = ["b", 2, 3];
	const Array a2 = ["a", 2, 3];
```

File: tests/test_array.h [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+	CHECK_MESSAGE(
+			a2 < a1,
+			"Array with a smaller item at index 1 should be smaller");
+}
+
+TEST_CASE("Array < Length") {
+	const Array a1 = [1]
+	const Array a2 = [1, 0]
```
Comment (Calinou) [inline tests/test_array.h:79]:
```suggestion
	const Array a1 = [1];
	const Array a2 = [1, 0];
```

File: tests/test_array.h [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+#ifndef TEST_ARRAY_H
+#define TEST_ARRAY_H
+
+#include "core/variant/array.h"
+#include "core/string/print_string.h"
+#include "tests/test_macros.h"
+
+#include "thirdparty/doctest/doctest.h"
+
+namespace TestArray {
```
Comment (Calinou) [inline tests/test_array.h:40]: There should be a blank line after the namespace declaration:

```suggestion
namespace TestArray {

```

File: tests/test_array.h [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	const Array a1 = [1]
+	const Array a2 = [1, 0]
+
+	CHECK_MESSAGE(
+			a1 < a2,
+			"Array with a shorter length should be smaller");
+}
+} // namespace TestArray
+
+#endif // TEST_Array_H
```
Comment (Calinou) [inline tests/test_array.h:87]:
```suggestion
#endif // TEST_ARRAY_H
```

File: tests/test_array.h [hunk 8]
```diff
@@ -0,0 +1,87 @@
+/*************************************************************************/
+/*  test_array.h                                                          */
```
Comment (Calinou) [inline tests/test_array.h:2]: Align the end of the comment with the lines above and below:

```suggestion
/*  test_array.h                                                         */
```

**Review by Calinou — CHANGES_REQUESTED** [review]

**Comment by akien-mga** [comment]
The tests never compiled and weren't fixed so closing.

The bug actually seems to be fixed in 4.0 already though the code is still the same. For 3.x I can still reproduce the issue so this might still be relevant to salvage, though I don't understand why this fix changes the behavior.

**Pull request closed without merging**